thrust/system/detail/generic/shuffle.inl

```diff
@@ -32,85 +32,98 @@ namespace system {
       state = do_round(state, i);
-    uint64_t res = state.B << w | state.A;
-    return res;
-  }
+    }
 
-private:
-  template <class URBG>
-  __host__ __device__ void init_state(URBG&& g)
-  {
-    thrust::uniform_int_distribution<uint32_t> dist(0, get_mask());
-    for( uint32_t i = 0; i < state_size; i++ )
-      S[i] = dist(g);
+    // Check we have the correct number of bits on each side
+    assert((state.left >> left_side_bits) == 0);
+    assert((state.right >> right_side_bits) == 0);
+
+    // Combine the left and right sides together to get result
+    return state.left << right_side_bits | state.right;
   }
 
+ private:
   // Find the nearest power of two
   __host__ __device__ uint64_t get_cipher_bits(uint64_t m) {
-    if(m == 0)
-      return 0;
     uint64_t i = 0;
-    m--;
     while (m != 0) {
       i++;
-      m >>= 2u;
+      m >>= 1;
     }
     return i;
   }
 
-  __host__ __device__ uint32_t get_mask() const
-  {
-    return (uint32_t)((1ull << (uint64_t)w) - 1ull);
+  // Equivalent to boost::hash_combine
+  __host__ __device__ size_t hash_combine(uint64_t lhs, uint64_t rhs) const {
+    lhs ^= rhs + 0x9e3779b9 + (lhs << 6) + (lhs >> 2);
+    return lhs;
   }
 
-  __host__ __device__ uint32_t rotl( uint32_t val, uint32_t amount ) const
-  {
-    const uint32_t amount_mod = amount % w;
-    return val << amount_mod | val >> (w-amount_mod);
+  // Round function, a 'pseudorandom function' who's output is indistinguishable
+  // from random for each key value input. This is not cryptographically secure
+  // but sufficient for generating permutations. 
+  __host__ __device__ uint32_t round_function(uint64_t value,
+                                              const uint64_t key) const {
+    uint64_t hash0 = thrust::random::taus88(value)();
+    uint64_t hash1 = thrust::random::ranlux48(value)();
+    return hash_combine(hash_combine(hash0, key), hash1) & left_side_mask;
   }
 
   __host__ __device__ round_state do_round(const round_state state,
                                            const uint64_t round) const {
-    uint32_t A = state.A;
-    uint32_t B = state.B;
-
-    A = ( rotl( A ^ B, B ) + S[ 2 * round + 2 ] ) & get_mask();
-    B = ( rotl( A ^ B, A ) + S[ 2 * round + 3 ] ) & get_mask();
-
-    return { A, B };
+    const uint32_t new_left = state.right & left_side_mask;
+    const uint32_t round_function_res =
+        state.left ^ round_function(state.right, key[round]);
+    if (right_side_bits != left_side_bits) {
+      // Upper bit of the old right becomes lower bit of new right if we have
+      // odd length feistel
+      const uint32_t new_right =
+          (round_function_res << 1ull) | state.right >> left_side_bits;
+      return {new_left, new_right};
+    }
+    return {new_left, round_function_res};
   }
 
-  static constexpr uint64_t state_size = 2 * num_rounds + 3;
-  const uint32_t w = 0;
-  uint32_t S[state_size];
+  static const uint64_t num_rounds = 16;
```

Comment:
Should this be constexpr? Im not sure the standards but generally its a good idea if you can isnt it? That way the object is smaller since it doesnt need the field (not that it matters really).

Reply:
I'll make it a template parameter again.